sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala

```diff
@@ -47,6 +47,18 @@ case class ShuffledHashJoinExec(
     "buildDataSize" -> SQLMetrics.createSizeMetric(sparkContext, "data size of build side"),
     "buildTime" -> SQLMetrics.createTimingMetric(sparkContext, "time to build hash map"))
 
+  override def outputPartitioning: Partitioning = joinType match {
```

Comment:
This is exactly the same as SMJ. Shall we create a common trait `ShuffleJoin` to put it?

Reply:
@cloud-fan - [there's an extra case for sort merge join to handle full outer join](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala#L81). I am thinking to handle all other join types in parent trait `ShuffleJoin`, and override `outputPartitioning` in `SortMergeJoinExec` to handle the extra `FullOuter`? What do you think?

But for me it's kind of weird that `ShuffleJoin` not handle  `FullOuter` as shuffled `FullOuter` join is one kind of `ShuffleJoin`. But if `ShuffleJoin` handles `ShuffleJoin`, it seems to be also weird that `ShuffledHashJoinExec` extends it.

Wondering what do you think? The change itself is easy. Thanks.

Reply:
Why does shuffle hash join not support `FullOuter`?

Reply:
> Why does shuffle hash join not support FullOuter?

@cloud-fan sorry if I miss anything, but [isn't this true now](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L196-L210)? Given current spark implementation for hash join, stream side looks up in build side hash map, it can handle non-matching keys from stream side if there's no match in build side hash map. But it cannot handle non-matching keys from build side, as there's no info persisted from stream side.

I feel an interesting followup could be to handle full outer join in shuffled hash join, where when looking up stream side keys from build side `HashedRelation`. Mark this info inside build side `HashedRelation`, and after reading all rows from stream side, output all non-matching rows from build side based on modified `HashedRelation`.

Reply:
Can we simply move `SortMergeJoinExec.outputPartitioning` to the parent trait? It works for `ShuffledHashJoinExec` as well, as the planner guarantees `ShuffledHashJoinExec.joinType` won't be `FullOuter`

Reply:
@cloud-fan I agree. Updated.